### DISCUSSION TITLE
커맨드 패턴을, 사용한 경험과 함께 알려 주세요.

### DISCUSSION THREAD

**Discussion by 3Juhwan** [2024-woowacourse-study/level-interview#94] [OOP]
.

**Comment by J-I-H-O** [comment]
커맨드패턴은 객체의 행위(메서드)를 클래스로 만들어 캡슐화하는 디자인 패턴입니다.
요청을 객체로 캡슐화하여 요청을 매개변수화합니다. 이는 요청을 발생시키는 객체와 이를 처리하는 객체를 분리하여 요청을 수신하는 객체가 실제 작업을 수행하는 객체에 대해 알 필요 없이 요청을 처리할 수 있도록 합니다.

미션을 해결하는데 급급해 새로운 패턴을 적용하기에는 부담이 있어 실제로 적용해보지는 않았습니다.

**Comment by 3Juhwan** [comment]
커맨드 패턴은 요청을 객체의 형태로 캡슐화하여 다양한 요청을 매개변수화하는 디자인 패턴입니다. 이 패턴은 요청을 발생시키는 객체와 요청을 처리하는 객체를 분리하여 상호작용하는 데 사용됩니다.

체스 미션에서는 커맨드 패턴을 적용했습니다. 여기서 커맨드는 뷰와 도메인 모두에 의존하는 형태로 비즈니스 로직을 실행하고, 그 결과를 표현하기 위해 뷰에 요청하는 역할을 담당했습니다. 이를 통해 Controller 코드가 단순화되었습니다. 긴 if문으로 작성되었던 코드를 외부 커맨드 객체로 분리함으로써 코드가 간결해졌습니다. 또한, 요구 사항으로 새로운 커맨드를 추가해야 한다면 새로운 구현체만 추가하여 완성할 수 있었습니다. 이렇게 함으로써 코드의 유지보수성이 향상되고, 확장성이 높아졌습니다.
